### PR TITLE
Refactor http server kind

### LINKED ISSUE #6017: Refactor WsHttpServerKind to Simplify Service Combinations and Enable Custom Services
### Describe the feature

- In our current implementation of the `WsHttpServerKind` enum, we handle various combinations of HTTP servers, including options with CORS, authentication, or both. This approach leads to a complex and somewhat rigid structure, which could be streamlined for better maintainability and flexibility.

- Refactoring the `WsHttpServerKind` enum to simplify the service combinations and, ideally, to allow the installation of custom services. This change would likely involve some API alterations and significant refactoring.

- We could explore the use of tower::Service and OptionLayer to create a more flexible and cleaner structure. This might allow for a more elegant way of stacking or combining services based on runtime configuration rather than hardcoding all possible combinations.

As Discussed with @mattsse  

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by supernovahs**
closes #6017 

Still figuring out how to simplify it . 
Opening for ref  and inputs if any .

### CODE REVIEW

File: crates/rpc/rpc-builder/src/lib.rs [hunk 1]
```diff
@@ -1918,18 +1922,22 @@ impl Default for WsHttpServers {
...
-    Plain(Server<Identity, RpcServerMetrics>),
-    /// Http server with cors
-    WithCors(Server<Stack<CorsLayer, Identity>, RpcServerMetrics>),
-    /// Http server with auth
-    WithAuth(Server<Stack<AuthLayer<JwtAuthValidator>, Identity>, RpcServerMetrics>),
-    /// Http server with cors and auth
-    WithCorsAuth(
-        Server<Stack<AuthLayer<JwtAuthValidator>, Stack<CorsLayer, Identity>>, RpcServerMetrics>,
-    ),
+    Server(Box<dyn Service<Request<Body>, Response = Response<Body>, Error = hyper::Error, Future = Pin<Box<dyn Future<Output = Result<Response<Body>, hyper::Error>>>>> + Send + Sync>),
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1940]: The type signature will be very big but we can flatten all of this into a bunch of `Either<Layer, Identiy>` which is what OptionLayer is:

https://docs.rs/tower/latest/tower/builder/struct.ServiceBuilder.html#method.option_layer

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc-builder/src/lib.rs [hunk 2]
```diff
@@ -1918,100 +1923,40 @@ impl Default for WsHttpServers {
...
-    Plain(Server<Identity, RpcServerMetrics>),
-    /// Http server with cors
-    WithCors(Server<Stack<CorsLayer, Identity>, RpcServerMetrics>),
-    /// Http server with auth
-    WithAuth(Server<Stack<AuthLayer<JwtAuthValidator>, Identity>, RpcServerMetrics>),
-    /// Http server with cors and auth
-    WithCorsAuth(
-        Server<Stack<AuthLayer<JwtAuthValidator>, Stack<CorsLayer, Identity>>, RpcServerMetrics>,
-    ),
+    Server(Server<tower::layer::util::Stack<tower::util::Either<reth_rpc::AuthLayer<reth_rpc::JwtAuthValidator>, tower::layer::util::Identity>, tower::layer::util::Stack<tower::util::Either<std::result::Result<tower_http::cors::CorsLayer, cors::CorsDomainError>, tower::layer::util::Identity>, tower::layer::util::Identity>>, metrics::RpcServerMetrics>),
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1928]: yeah this is a bit horrible, but better than before,

can convert this to a struct now

File: crates/rpc/rpc-builder/src/lib.rs [hunk 3]
```diff
@@ -1918,100 +1927,77 @@ impl Default for WsHttpServers {
...
-        match self {
-            WsHttpServerKind::Plain(server) => server.start(module),
-            WsHttpServerKind::WithCors(server) => server.start(module),
-            WsHttpServerKind::WithAuth(server) => server.start(module),
-            WsHttpServerKind::WithCorsAuth(server) => server.start(module),
-        }
-    }
+    // async fn start( self, module: RpcModule<()>) -> ServerHandle {
+
+    //   #6133  How to call start() ?
```
Comment (supernovahs) [inline crates/rpc/rpc-builder/src/lib.rs:1968]: figuring out this

File: crates/rpc/rpc-builder/src/lib.rs [hunk 4]
```diff
@@ -1918,100 +1927,77 @@ impl Default for WsHttpServers {
...
-    WithAuth(Server<Stack<AuthLayer<JwtAuthValidator>, Identity>, RpcServerMetrics>),
-    /// Http server with cors and auth
-    WithCorsAuth(
-        Server<Stack<AuthLayer<JwtAuthValidator>, Stack<CorsLayer, Identity>>, RpcServerMetrics>,
-    ),
+pub struct WsHttpServerKind {
+    server: ServerOriginal,
 }
 
+type EitherServer = Either<
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1934]: no type aliases, we want struct wrapper for this

**Comment by supernovahs** [comment]
Any way to resolve this error? 
```
error[E0599]: the method `start` exists for struct `Server<Stack<Either<AuthLayer<JwtAuthValidator>, Identity>, ...>, ...>`, but its trait bounds were not satisfied
    --> crates/rpc/rpc-builder/src/lib.rs:1960:35
     |
1960 |                 Either::A(a) => a.start(module).await, 
     |                                   ^^^^^ method cannot be called due to unsatisfied trait bounds
     |
    ::: /Users/supernovahs/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tower-0.4.13/src/util/either.rs:22:1
     |
22   | pub enum Either<A, B> {
     | --------------------- doesn't satisfy `_: Layer<Either<AuthService<TowerService<RpcServerMetrics>, JwtAuthValidator>, TowerService<RpcServerMetrics>>>`
     |
    ::: /Users/supernovahs/.rustup/toolchains/nightly-2023-12-21-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/result.rs:502:1
     |
502  | pub enum Result<T, E> {
     | --------------------- doesn't satisfy `_: Layer<Either<AuthService<TowerService<RpcServerMetrics>, JwtAuthValidator>, TowerService<RpcServerMetrics>>>`
     |
    ::: /Users/supernovahs/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tower-layer-0.3.2/src/stack.rs:6:1
     |
6    | pub struct Stack<Inner, Outer> {
     | ------------------------------
     | |
     | doesn't satisfy `_: Layer<Either<AuthService<TowerService<RpcServerMetrics>, JwtAuthValidator>, TowerService<RpcServerMetrics>>>`
     | doesn't satisfy `_: Layer<TowerService<RpcServerMetrics>>`
     |
     = note: the full type name has been written to '/Users/supernovahs/Desktop/reth/target/debug/deps/reth_rpc_builder-a7d87878da363f83.long-type-2193867962893713952.txt'
     = note: the following trait bounds were not satisfied:
             `Stack<tower::util::Either<AuthLayer<JwtAuthValidator>, Identity>, Stack<tower::util::Either<std::result::Result<CorsLayer, CorsDomainError>, Identity>, Identity>>: Layer<jsonrpsee::jsonrpsee_server::server::TowerService<RpcServerMetrics>>`
             `Stack<tower::util::Either<std::result::Result<CorsLayer, CorsDomainError>, Identity>, Identity>: Layer<tower::util::Either<reth_rpc::layers::auth_layer::AuthService<jsonrpsee::jsonrpsee_server::server::TowerService<RpcServerMetrics>, JwtAuthValidator>, jsonrpsee::jsonrpsee_server::server::TowerService<RpcServerMetrics>>>`
             `tower::util::Either<std::result::Result<CorsLayer, CorsDomainError>, Identity>: Layer<tower::util::Either<reth_rpc::layers::auth_layer::AuthService<jsonrpsee::jsonrpsee_server::server::TowerService<RpcServerMetrics>, JwtAuthValidator>, jsonrpsee::jsonrpsee_server::server::TowerService<RpcServerMetrics>>>`
             `std::result::Result<CorsLayer, CorsDomainError>: Layer<tower::util::Either<reth_rpc::layers::auth_layer::AuthService<jsonrpsee::jsonrpsee_server::server::TowerService<RpcServerMetrics>, JwtAuthValidator>, jsonrpsee::jsonrpsee_server::server::TowerService<RpcServerMetrics>>>`

```

**Pull request closed without merging**